File: apps/clojure/design-pattern/test/design_pattern/command_test.clj
(ns design-pattern.command-test
  (:require [clojure.test :refer :all]
            [design-pattern.command :refer :all]))

(deftest create-and-execute-command-test
  (testing "コマンドを作成して実行する"
    (let [state (atom {:files {}})
          cmd   (create-file-command (:files state) "test.txt" "hello")]
      ;; state を直接使う
      (let [fs  (atom {})
            cmd (create-file-command fs "test.txt" "hello")]
        (execute cmd)
        (is (= {"test.txt" "hello"} @fs))))))

(deftest undo-command-test
  (testing "コマンドの取り消しができる"
    (let [fs  (atom {})
          cmd (create-file-command fs "test.txt" "hello")]
      (execute cmd)
      (is (= {"test.txt" "hello"} @fs))
      (undo cmd)
      (is (= {} @fs)))))

(deftest delete-and-undo-test
  (testing "削除コマンドの実行と取り消し"
    (let [fs      (atom {"readme.md" "# README"})
          del-cmd (delete-file-command fs "readme.md")]
      (execute del-cmd)
      (is (nil? (get @fs "readme.md")))
      (undo del-cmd)
      (is (= "# README" (get @fs "readme.md"))))))

(deftest composite-command-test
  (testing "コンポジットコマンドで複数操作をまとめて実行する"
    (let [fs   (atom {})
          cmds [(create-file-command fs "a.txt" "aaa")
                (create-file-command fs "b.txt" "bbb")]
          comp-cmd (composite-command "Create files" cmds)]
      (execute comp-cmd)
      (is (= {"a.txt" "aaa" "b.txt" "bbb"} @fs))
      (undo comp-cmd)
      (is (= {} @fs)))))

(deftest command-history-test
  (testing "コマンド履歴でundo ができる"
    (let [fs      (atom {})
          history (create-history)]
      (execute-with-history! history (create-file-command fs "x.txt" "xxx"))
      (execute-with-history! history (create-file-command fs "y.txt" "yyy"))
      (is (= {"x.txt" "xxx" "y.txt" "yyy"} @fs))
      (undo-last! history)
      (is (= {"x.txt" "xxx"} @fs))
      (undo-last! history)
      (is (= {} @fs)))))

File: apps/elixir/design-pattern/test/design_pattern/command_test.exs
defmodule DesignPattern.CommandTest do
  use ExUnit.Case
  alias DesignPattern.Command

  test "コマンドを実行できる" do
    cmd = Command.new("add", fn n -> n + 10 end, fn n -> n - 10 end)
    assert Command.execute(cmd, 5) == 15
  end

  test "コマンドを取り消せる" do
    cmd = Command.new("add", fn n -> n + 10 end, fn n -> n - 10 end)
    assert Command.undo(cmd, 15) == 5
  end

  test "コマンド履歴で実行と取り消しができる" do
    cmd1 = Command.new("double", fn n -> n * 2 end, fn n -> div(n, 2) end)
    cmd2 = Command.new("add5", fn n -> n + 5 end, fn n -> n - 5 end)

    history =
      %{Command.new_history() | state: 10}
      |> Command.execute_with_history(cmd1)
      |> Command.execute_with_history(cmd2)

    assert history.state == 25

    undone = Command.undo_last(history)
    assert undone.state == 20
  end

  test "空の履歴で undo しても安全" do
    history = Command.new_history()
    assert Command.undo_last(history) == history
  end

  test "マクロコマンドで複数操作を一括実行する" do
    cmd1 = Command.new("a", fn n -> n + 1 end, fn n -> n - 1 end)
    cmd2 = Command.new("b", fn n -> n * 3 end, fn n -> div(n, 3) end)
    macro = Command.macro("macro", [cmd1, cmd2])

    assert Command.execute(macro, 2) == 9
    assert Command.undo(macro, 9) == 2
  end
end
